### PR TITLE
Error handling

### PR DESCRIPTION
**Pull request by oleksiikulyk-tomtom**

### CODE REVIEW

File: components/Chat/Chat.tsx [hunk 1]
```diff
@@ -112,16 +114,61 @@ export const Chat = memo(({stopConversationRef}: Props) => {
...
+            if (error.errorType === "openai_auth_error") {
+              toast.error("Invalid API Key. Please enter the correct Azure OpenAI key in left menu bar of Chatty.", {
+                duration: TOAST_DURATION_MS
+              })
+              return
+            }
+
+            if (error.errorType === "context_length_exceeded") {
+              toast.error(
+                `This model's maximum context length is exceeded (${error.limit} tokens but ${error.requested} tokens requested). Please reduce the number of tokens in the request`,
```
Comment (RijnBuve-TomTom) [inline components/Chat/Chat.tsx:136]: Missing '.' at end of message.

File: components/Chat/ChatInput.tsx [hunk 2]
```diff
@@ -340,6 +339,7 @@ Please remove some messages from the conversation, or simply clear all previous
             onKeyDown={handleKeyDown}
           />
           <button
+            aria-label="Send Message"
```
Comment (RijnBuve-TomTom) [inline components/Chat/ChatInput.tsx:342]: Lowercase 'm': "Send messages" (we don't do 'Capital Casing' in Chatty, just 'Sentence casing').

File: components/UnlockCode/useUnlockCodeInterceptor.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+          Authorization: `Bearer ${code}`
+        }
+      }
+    },
+    response: async ({response}) => {
+      if (response.status === 401) {
+        const clonedResponse = response.clone()
+        if (clonedResponse.headers.get("content-type")?.indexOf("text/plain") !== -1) {
+          const error = await clonedResponse.text()
+          if (error.includes("You are not authorized to use the service. Check your Unlock code.")) {
```
Comment (RijnBuve-TomTom) [inline components/UnlockCode/useUnlockCodeInterceptor.ts:37]: I'd make this more resilient to a change in the text. Just check for for "not authorized", not the entire sentence.

File: utils/server/openAiClient.ts [hunk 4]
```diff
@@ -83,11 +112,36 @@ export const ChatCompletionStream = async (
...
+
+    if (response.status === 429 && result.error) {
+      const match = result.error.message.match(/retry after (\d+) seconds/)
+      const retryAfter = match ? parseInt(match[1]) : undefined
+      throw new OpenAIRateLimited(result.error.message, retryAfter)
+    }
+
+    if (result.error && result.error.code === "context_length_exceeded") {
+      const match = result.error.message.match(
+        /maximum context length is (\d+) tokens. However, you requested (\d+) tokens/
```
Comment (RijnBuve-TomTom) [inline utils/server/openAiClient.ts:131]: Perhaps use the regex to make more resilient to change.
`/max.*length.* (\d+) tokens.*requested (\d+\) tokens/`

File: utils/server/openAiClient.ts [hunk 5]
```diff
@@ -83,11 +112,36 @@ export const ChatCompletionStream = async (
...
+    if (response.status === 401) {
+      if (result.error) {
+        throw new OpenAIAuthError(result.error.message)
+      }
+
+      throw new OpenAIAuthError(result.message)
+    }
+
+    if (response.status === 429 && result.error) {
+      const match = result.error.message.match(/retry after (\d+) seconds/)
```
Comment (RijnBuve-TomTom) [inline utils/server/openAiClient.ts:124]: Use the regex to make more resilient to change:
`/retry.* \(\d+) sec/`

File: utils/server/tiktoken.ts [hunk 6]
```diff
@@ -21,19 +22,45 @@ export async function prepareMessagesToSend(
...
+  const [messagesToSend, requiredTokens] = reduceMessagesToSend(
+    tokenLimit,
+    maxReplyTokens,
+    prompt,
+    messages,
+    modelId,
+    encoding
+  )
+
+  if (requiredTokens > tokenLimit) {
```
Comment (RijnBuve-TomTom) [inline utils/server/tiktoken.ts:34]: See Slack discussion. Not sure if this is sufficiently user friendly. It's code-simple, but user unfriendly. But maybe we should leave it like this and add an auto-summarize and restart conversation instead (see issue https://github.com/rijnb/chatty-server/issues/53).

**Review by RijnBuve-TomTom — CHANGES_REQUESTED** [review]

File: utils/server/tiktoken.ts (hunk 6 above)
Comment (oleksiikulyk-tomtom) [reply utils/server/tiktoken.ts:34]: this is kind of a safety net, when even system prompt + user prompt (first message) don't fit.
skipping messages still happening [reduceMessagesToSend](https://github.com/rijnb/chatty-server/pull/76/files#diff-0a44d093b1e2a00d0b358a8e7de28b0594346aeef6c6fcd37806a21a5583c5e9R41-R43)

**Review by RijnBuve-TomTom — APPROVED** [review]

**Pull request merged**